Implementation of the `/api/days` method, some documentation updates

kolayne opened the pull request:
Relates to #19 

So sorry I make backwards-incompatible updates of the docs, while front-end is being built on top(?) of them! :'(
Apparently, I didn't think deep enough when drawing the documentation up. There are mostly clarifications or new details added, but there are currently at least two incompatible fixes: I replaced the `400 Bad Request` status code with `412 Precondition Failed` for some cases and updated date format for one method. So, @tanya-kta and anyone else who needs the documentation, please, take a look at the changes and tell me if there is something unclear about what I've added (it's a bit late now, so I might have written some nonsense or explained something too complicatedly)

Welcome the new method for adding new days!

Is blocked by #29 because of the following:
https://github.com/LifeCalendarTeam/life_calendar/blob/b31770c3098f3246447f761bb475c4a6cf704a4e/src/webhandlers.go#L118-L120

kolayne committed:
Updated api docs: use 412 instead of 400 in `/api/days` in one of the cases

Also added some details on 400/412 responses to the general methods description above

kolayne committed:
Multiple updates to the API docs

- Added a clarification of the status selection when there are multiple errors
- [BACKWARD COMPATIBILITY BREAKING] Updated date format in the docs of `/api/days`
- Added much more details about errors which may occur in `/api/days`
- Added the `error_type` field to the response of `/api/days`, helping to identify the error occurred during a request processing based on the response
- Moved response description to the "Response" subsection of the docs of `/api/days`

kolayne committed:
Added a draft of `/api/days` method

Wow, it's so big, there are so many TODOs... I haven't even checked that this is working yet

kolayne commented on src/webhandlers.go:
```diff
@@ -113,6 +113,57 @@ func getAverageColor(proportionsAndColors []proportionAndColor) ([3]int, error)
+		return
+	}
+
+	// TODO: the two lines below are formatted automatically this way. I wonder if it is possible to write it better
+	if len(r.Form["activity_type"]) != len(r.Form["activity_proportion"]) ||
+		len(r.Form["emotion_type"]) != len(r.Form["emotion_proportion"]) {
+
+		js, err := json.Marshal(map[string]interface{}{"ok": false,
```
@DarkSquirrelComes, I am planning to request your complete review on the PR when I finish it, but I have a question so far. I hate how these lines look together (especially if I remove the blank line). Is there a nice way to fix it? Our hard limit is 120 characters per line (including the indentation, my tab size is 4)

I would also be thankful if you have checked other TODOs added in this PR and tell me if there is something you disagree with or know a better way for. Thank you!

DarkSquirrelComes replied:
This autoformat seems good-looking to me. But I don't really understand, why you compare lens of this values?

kolayne replied:
> This autoformat seems good-looking to me.

Then should I level the empty line there? Look what happens if I remove it:
```Golang
	// TODO: the two lines below are formatted automatically this way. I wonder if it is possible to write it better
	if len(r.Form["activity_type"]) != len(r.Form["activity_proportion"]) ||
		len(r.Form["emotion_type"]) != len(r.Form["emotion_proportion"]) {
		js, err := json.Marshal(map[string]interface{}{"ok": false,
```
I can't understand where the condition of this `if` ends and the body begins based on the indentation!


> But I don't really understand, why you compare lens of this values?

Each activity/emotion here is described with one type and one proportion. I ensure there exists a type for each proportion and a proportion for each value. You can see more details on this method in [the documentation](https://github.com/LifeCalendarTeam/life_calendar/blob/api_days/docs/api/README.md#apidays)

kolayne committed:
Fixed some issues found by Codacy

Still don't fully understand the rules of Markdown indentation... Hope, this is correct
(oops, accidentally fixed an older issue...)

kolayne committed:
A bit more specific constraint on proportions in `/api/days`

VladimirKoryakin committed:
Added transaction in `HandleAPIDays`

Adding new days is now done via one transaction

Co-authored-by: Nikolay Nechaev <Nikolay_Nechaev@mail.ru>

VladimirKoryakin committed:
Added check constraint on `activities_and_emotions` table

Only activities and emotions which belong to user X can now be added to day  owned by user X

Co-authored-by: Nikolay Nechaev <Nikolay_Nechaev@mail.ru>

kolayne committed:
`HandleAPIDays`: PgSQL API + error handling

Updated the way to retrieve the id of the added day, added duplicated day error handling

kolayne committed:
SQL scheme: added the missing `UNIQUE` constraint on `days`

kolayne committed:
Minor (formatting)

kolayne committed:
SQL scheme: gave check constraint a name to make the error handling code more specific

Also added a comment about the constraints names to the scheme file

kolayne committed:
`HandleAPIDays`: added incorrect type/proportion errors handlers

kolayne committed:
API docs: add `bad_ae_type` (optional) field to the response of `/api/days`

This is not implemented yet

kolayne committed:
Made `MaxProportion` an `int` constant, updated proportion value check in `HandleAPIDays`

Compare the value to the `MaxProportion` constant, not to the number `100`

kolayne committed:
Merge branch 'master' into api_days

# Conflicts:
#	database_scheme.sql
#	src/webhandlers.go

kolayne committed:
Updated `writeJSON` to reduce copy-pasts even more

kolayne committed:
Minor style

kolayne committed:
Documented the `/api/days` method behavior when an activity/emotion is duplicated

kolayne committed:
Implemented activities/emotions duplication protection in /api/days

kolayne committed:
Added `bad_ae_type` to return values of the /api/days method

kolayne committed:
Replaced "type id"s with "type"s in the error messages

"Type id" is an object in the SQL scheme, which is called "type" in the API docs

kolayne commented:
That's pretty much it. Waiting for @VladimirKoryakin to resolve #29.

I would also like the reviewers (@DarkSquirrelComes, @tanya-kta) to pay additional attention to comparing the documented and the actual behaviours of the method in this PR! Please, ensure all the errors expected to be returned __are__ returned in all the appropriate cases

kolayne commented:
Note: I want both two reviews on this

kolayne commented:
One more thing: the database schema is invalid because PostgreSQL does not support `CHECK` constraints that reference data that is not in the row being inserted or updated ([docs](https://www.postgresql.org/docs/current/ddl-constraints.html)). This **must** be fixed before the PR can be merged ~~(if this will ever happen)~~
